modules/objdetect/include/opencv2/objdetect/aruco_board.hpp

```diff
@@ -220,6 +220,16 @@ class CV_EXPORTS_W CharucoBoard : public Board {
      */
     CV_WRAP bool checkCharucoCornersCollinear(InputArray charucoIds) const;
 
+    /** @brief Given a board configuration and a set of detected markers, returns the corresponding
+     * image points and object points to call solvePnP
+     *
+     * @param detectedCorners List of detected marker corners of the board.
+     * @param detectedIds List of identifiers for each marker.
+     * @param objPoints Vector of vectors of board marker points in the board coordinate space.
+     * @param imgPoints Vector of vectors of the projections of board marker corner points.
+     */
+    CV_WRAP void matchImagePoints(InputArrayOfArrays detectedCorners, InputArray detectedIds,
+                                  OutputArray objPoints, OutputArray imgPoints) const override;
 protected:
     struct CharucoImpl;
```

Comment:
Do you suggest make PImpl fields as private and prevent class inheritors using this field? In ImageCollection(imgcodecs), Model (dnn), IntelligentScissorsMB classes PImpl fields set as protected and can be used by the class's inheritors.

Reply:
> private

Where is it suggested?

---

They must reuse the **single** "impl" field instead of adding a new one.

See `Model` (dnn) inheritance pattern.

Reply:
fields `GridImpl` and `CharucoImpl` were removed from GridBoard, CharucoBoard

Reply:
So, change in `virtual void matchImagePoints()` is not needed anymore.

---

BTW, too many unnecessary `Ptr<>` in implementation, especially without NULL checks.
Reference should be used instead.

Reply:
change most `Ptr<>` to reference

Reply:
@alalek, do you want me to remove the virtuality for the function `matchImagePoints()` and `generateImage()`? Do you want me to use virtuality into Pimpl implementation?

Reply:
I want this kind of code to work correctly:
```
    cv::Ptr<cv::aruco::Board> board = cv::aruco::CharucoBoard::create(3, 3);
    std::vector<cv::Point2f> a1;
    std::vector<int> a2;
    std::vector<cv::Point2f> a3;
    std::vector<cv::Point2f> a4;
    board->matchImagePoints(a1, a2, a3, a4);
```
without `virtual` in `matchImagePoints()` I need to create func `virtual void pimplMatchImagePoints` and call this from main classes

Should I do it?

Reply:
This done